text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,269 @@
+- Feature Name: `expose-fn-type`
+- Start Date: 2023-08-20
+- RFC PR: [rust-lang/rfcs#3476](https://github.com/rust-lang/rfcs/pull/3476)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+This exposes the function type of a function item to the user.
+
+# Motivation
+[motivation]: #motivation
+
+### DasLixou
+
+I was trying to make something similar to bevy's system functions. And for safety reasons, they check for conflicts between SystemParams, so that a function requiring `Res<A>` and `ResMut<A>` [panic](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/system/system_param.rs#L421).
+
+Then after I heard about axum's [`#[debug_handler]`](https://docs.rs/axum/latest/axum/attr.debug_handler.html) I wanted to do something similar to my copy of bevy systems, so that I get compile time errors when there is a conflict. I wanted even more, I wanted to force the user to mark the function with a specific proc attribute macro in order to make it possible to pass it into my code and call itself a system.
+
+For that, I would need to mark the type behind the function item, for example, with a trait.
+
+### madsmtm
+
+In Swift, some functions have an associated selector that you can access with [`#selector`](https://developer.apple.com/documentation/swift/using-objective-c-runtime-features-in-swift).
+
+In my crate `objc2`, it would be immensely beautiful (and useful) to be able to do something similar, e.g. access a function's selector using something like `MyClass::my_function::Selector` or `selector(MyClass::my_function)`, instead of having to know the selector name (which might be something completely different than the function name).
+
+# Terminology
+
+I'll may shorten `function` to `fn` sometimes.
+
+- **function pointer**: pointer type with the type syntax `fn(?) -> ?` directly pointing at a function, not the type implementing the `Fn[Once/Mut](?) -> ?` traits.
+- **function item** (or just function): a declared function in code. free-standing or associated to a type.
+- **function group**: many non-specific functions with the same signature (params, return type, etc.)
+- **function trait(s)**: the `Fn[Once/Mut](?) -> ?` traits
+- **function type**: the type behind a function, which also implements the function traits.
+- **fixed type**: directly named type, no generic / `impl Trait`.
+- **describe the function type**: write `fn(..) -> ? name` instead of just `fn name`.
+
+# Guide-level explanation
```

Comment:
Although I agree that this is where we want to end up, I think that this RFC is focusing too much on the `impl Trait for [FunctionType]` use-case.

Maybe it would make sense to restrict the RFC to just figuring out the syntax for `fn` items? Or make a new RFC for just the syntax? I would be willing to drive that effort, if you think so?

Implementing traits for functions have many other nuances, I'll just name a few that I don't think have been explored enough in this RFC:
- Coherence: what traits should you be allowed to implement for your function? Can I implement `IntoIterator` for my function? What if 10 years down the road the language wants to start implementing `IntoIterator` for function items of the form `fn() -> impl Iterator`?
- How might it affect inference rules if you implement a trait for your function, where the trait is also implemented for `fn` pointers.

Reply:
I tried to find some other examples of when you'd want to talk about the function item type instead of just using the function pointer, and I honestly couldn't really find a compelling example, which is supposedly why it doesn't exist yet.

Reply:
@madsmtm good point with the fn thing! I honestly wonder how that would integrate with generator functions...

Reply:
Note to self: The standard library uses the macro `impl_fn_for_zst!` to support exactly the use-case of naming a function item, because they store it in generic helper structs, and would like that storage to be zero-cost.